kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java

```diff
@@ -0,0 +1,417 @@
+        for (int i = 0; i < virtualCluster.getNumberOfBrokerEndpointsToPrebind(); i++) {
+            initialBindings.put(virtualCluster.getBrokerAddress(i), i);
+        }
+
+        var endpointFutures = initialBindings.entrySet().stream().map(e -> {
+            var hp = e.getKey();
+            var nodeId = e.getValue();
+            var key = Endpoint.createEndpoint(bindingAddress, hp.port(), tls);
+            VirtualClusterBinding virtualClusterBinding = VirtualClusterBinding.createBinding(virtualCluster, nodeId);
+            return registerBinding(key, hp.host(), virtualClusterBinding).toCompletableFuture();
+            // TODO cache endpoints when future completes
+        }).toList();
+
+        var unused = allOfFutures(endpointFutures).whenComplete((u, t) -> {
+            var future = vcr.registrationStage.toCompletableFuture();
+            if (t != null) {
+                // Try to roll back any bindings that were successfully made
+                deregisterBinding(virtualCluster, (vcb) -> vcb.virtualCluster().equals(virtualCluster))
+                        .handle((u1, t1) -> {
+                            if (t1 != null) {
+                                LOGGER.warn("Registration error", t);
+                                LOGGER.warn("Secondary error occurred whilst handling a previous registration error: {}", t.getMessage(), t1);
+                            }
+                            registeredVirtualClusters.remove(virtualCluster);
+                            future.completeExceptionally(t);
+                            return null;
+                        });
+            }
+            else {
+                try {
+                    future.complete(endpointFutures.get(0).get());
+                }
+                catch (Throwable t1) {
+                    future.completeExceptionally(t1);
+                }
+            }
+        });
+
+        return vcr.registrationStage();
+    }
+
+    /**
+     * De-registers a virtual cluster from the registry, removing all existing endpoint bindings, and
+     * closing any listening sockets that are no longer required. This operation returns a {@link CompletionStage<Endpoint>}.
+     * The completion stage will complete once any necessary network unbind operations are completed.
+     *
+     * @param virtualCluster virtual cluster to be deregistered.
+     * @return completion stage that will complete after registration is finished.
+     */
+    public CompletionStage<Void> deregisterVirtualCluster(VirtualCluster virtualCluster) {
+        Objects.requireNonNull(virtualCluster, "virtualCluster cannot be null");
+
+        var deregisterFuture = new CompletableFuture<Void>();
+        var vcr = registeredVirtualClusters.get(virtualCluster);
+        if (vcr == null) {
+            // cluster not currently registered
+            deregisterFuture.complete(null);
+            return deregisterFuture;
+        }
```

Comment:
nit: I'd prefer to use `registeredVirtualClusters.containKey(virtualCluster)`.

I think it describes the intent better. I'm also wondering if it would be cleaner to return a completed future rather than a single `deregisterFuture`

Reply:
you just mean `CompletableFuture.completedFuture(null)`, right?

Reply:
The code needs the value of `vcr` a couple of lines below. I've made the other change.